telethon/full_sync.py

```diff
@@ -0,0 +1,156 @@
+    try:
+        res = await af
+        cf.set_result(res)
+    except Exception as e:
+        cf.set_exception(e)
+
+
+def _sync_result(loop, x):
+    f = futures.Future()
+    loop.call_soon_threadsafe(asyncio.ensure_future, _proxy_future(x, f))
+    return f.result()
+
+
+def _syncify_coro(t, method_name, loop, thread_name):
+    method = getattr(t, method_name)
+
+    @functools.wraps(method)
+    def syncified(*args, **kwargs):
+        coro = method(*args, **kwargs)
+        return (
+            coro if threading.current_thread().name == thread_name
+            else _sync_result(loop, coro)
+        )
+
+    setattr(t, method_name, syncified)
+
+
+class _SyncGen:
+    def __init__(self, loop, gen):
+        self.loop = loop
+        self.gen = gen
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return _sync_result(self.loop, self.gen.__anext__())
+        except StopAsyncIteration:
+            raise StopIteration from None
+
+
+def _syncify_gen(t, method_name, loop, thread_name):
+    method = getattr(t, method_name)
+
+    @functools.wraps(method)
+    def syncified(*args, **kwargs):
+        coro = method(*args, **kwargs)
+        return (
+            coro if threading.current_thread().name == thread_name
+            else _SyncGen(loop, coro)
+        )
+
+    setattr(t, method_name, syncified)
+
+
+def _syncify(*types, loop, thread_name):
+    for t in types:
+        for method_name in dir(t):
+            if not method_name.startswith('_') or method_name == '__call__':
+                if inspect.iscoroutinefunction(getattr(t, method_name)):
+                    _syncify_coro(t, method_name, loop, thread_name)
+                elif isasyncgenfunction(getattr(t, method_name)):
+                    _syncify_gen(t, method_name, loop, thread_name)
+
+
+__asyncthread = None
+
+
+def enable(loop=None, thread_name="__telethon_async_thread__"):
+    global __asyncthread
+    if __asyncthread is not None:
+        raise ValueError("full_sync has already been enabled.")
+
+    if not loop:
+        loop = asyncio.get_event_loop()
+    old_init = TelegramClient.__init__
+
+    @functools.wraps(old_init)
+    def new_init(*args, **kwargs):
+        kwargs['loop'] = loop
+        return old_init(*args, **kwargs)
+
+    TelegramClient.__init__ = new_init
+
+    _syncify(TelegramClient, Draft, Dialog, MessageButton, ChatGetter,
+             SenderGetter, Forward, Message, InlineResult, Conversation,
+             loop=loop, thread_name=thread_name)
+    _syncify_coro(TelegramClient, "start", loop, thread_name)
+
+    old_add_event_handler = TelegramClient.add_event_handler
+    old_remove_event_handler = TelegramClient.remove_event_handler
+    proxied_event_handlers = {}
+
+    @functools.wraps(old_add_event_handler)
+    def add_proxied_event_handler(self, callback, event_type=None):
+        async def _proxy(event):
+            h_t = threading.Thread(target=callback, args=(event,))
```

Comment:
Why not `loop.run_in_executor`, with a `ThreadPoolExecutor`? Spawning an unbounded number of threads is rather unhealthy and not performance-friendly.